apps/nextra/pages/en/build/smart-contracts/book/functions.mdx

```diff
@@ -734,16 +733,155 @@ The type of the `self` argument can be a struct or an immutable or mutable refer
 
 ```move
-module 0x42::example {
-    fun bar() {
-        let v = vector[1, 2, 3];
-        let e1 = v.remove(0); // type params inferred for `remove<T>`
-        assert!(e1 == 1);
-        let e2 = v.remove::<u8>(0); // type params explicitly specified
-        assert!(e2 == 2);
-    }
+struct S has key { 
+  func: |u64| bool /* has store */  // not needed since inherited
+}
+```
+
+### Operations on Functions
+
+A function is evaluated by providing the corresponding number of parameters, similar as when calling a named function. When a function is evaluated, the function value is *consumed*. Hence if the value needs to be evaluated multiple times, it's type must have the `copy` ability:
+
+```move
+let f: |u64|bool has copy = ...;
+assert!(f(1) == f(2))
+```
+
+Function values support equality and ordering. Note that those relations are based on the name of the underlying function behind a runtime value, and do not reflect semantic equivalence.
+
+
+### Function Type Wrappers
+
+Function types, specifically if they come together with abilities, can be verbose, and if the same type of function type is used many times in the code, repetitive. For this purpose, Move recognizes struct wrappers around function types as a special case. They can be used to effectively create named function types:
+
+```move
+struct Predicate<T>(|T|bool) has copy;
+```
+
+Move supports this feature by automatically converting function values into the wrapper type and vice versa. Examples:
+
+```move
+fun is_true(op: Predicate<u64>, param: u64) {
+    // Automatically converts into wrapped function and evaluates
+    op(param)
+}
+
+fun is_true_on_value(f: |u64|bool has copy): bool {
+    // Automatically converts into function wrapper
+    is_true(f, 42)
+}
+```
+
+### Denoting Function Values
+
+Function values can be constructed by directly using a function name. The resulting function type is derived from the signature of the underlying function, with abilities `copy+drop`. If the function is public, those function values have the `store` ability as well:
+
+```move
+public fun is_even(x: u64): bool { x % 2 == 0 }
+fun is_odd(x: u64): bool { x % 2 == 1 }
+...
+let f: |u64|bool has copy+drop+store = is_even;
+let g: |u64|bool has copy+drop = is_odd;
+```
+
+A public function is required to build a storable function value because it needs to be guaranteed that the underlying function persists so the function value can be safely restored from storage at any point in the future. However, code upgrade may change the underlying implementation of the function, while its signature is persistent. 
+
+If a function should storable which is non-public, an attribute can be used to mark this function as persistent, with the same upgrade behavior than public functions. This avoids exposing such a function outside of a package which can be a security risk:
```

Comment:
This sentence is off: "If a function should storable which is non-public" and "with the same upgrade behavior than public functions" do not make sense?